prql-compiler/src/sql/gen_expr.rs

```diff
@@ -76,132 +76,230 @@ pub(super) fn translate_expr(expr: Expr, ctx: &mut Context) -> Result<sql_ast::E
-        over: None,
-        distinct: false,
-        special: false,
-    })))
+fn collect_concat_args(expr: &Expr) -> Vec<&Expr> {
+    match &expr.kind {
+        ExprKind::BuiltInFunction { name, args } if name == "std.concat" => {
+            args.iter().flat_map(collect_concat_args).collect()
+        }
+        _ => vec![expr],
+    }
 }
 
-fn try_unpack_concat(expr: Expr) -> Result<Result<Vec<Expr>, Expr>> {
-    let Some((_, _)) = try_unpack(&expr, [STD_CONCAT])? else {
-        return Ok(Err(expr));
-    };
-    let [left, right] = unpack(expr, STD_CONCAT);
+/// Translate expr into a BETWEEN statement if possible, otherwise returns the expr unchanged.
+fn try_into_between(expr: Expr, ctx: &mut Context) -> Result<Option<sql_ast::Expr>, anyhow::Error> {
+    if let ExprKind::BuiltInFunction { name, args } = &expr.kind {
+        if name == "std.and" && args.len() == 2 {
+            let (a, b) = (&args[0], &args[1]);
+
+            if let (
+                ExprKind::BuiltInFunction {
+                    name: a_name,
+                    args: a_args,
+                },
+                ExprKind::BuiltInFunction {
+                    name: b_name,
+                    args: b_args,
+                },
+            ) = (&a.kind, &b.kind)
+            {
+                if a_name == "std.gte"
+                    && b_name == "std.lte"
+                    && a_args.len() == 2
+                    && b_args.len() == 2
+                {
+                    let (a_l, a_r) = (&a_args[0], &a_args[1]);
+                    let (b_l, b_r) = (&b_args[0], &b_args[1]);
+
+                    // We need for the values on each arm to be the same; e.g. x
+                    // > 3 and x < 5
+                    if a_l == b_l {
+                        return Ok(Some(sql_ast::Expr::Between {
+                            expr: translate_operand(a_l.clone(), 0, false, ctx)?,
+                            negated: false,
+                            low: translate_operand(a_r.clone(), 0, false, ctx)?,
+                            high: translate_operand(b_r.clone(), 0, false, ctx)?,
+                        }));
+                    }
+                }
+            }
+        }
+    }
+    Ok(None)
+}
 
-    let mut args = match try_unpack_concat(left)? {
-        Ok(args) => args,
-        Err(left) => vec![left],
-    };
-    args.push(right);
-    Ok(Ok(args))
+fn operator_from_name(name: &str) -> Option<BinaryOperator> {
+    use BinaryOperator::*;
+    match name {
+        "std.mul" => Some(Multiply),
+        "std.div" => Some(Divide),
+        "std.mod" => Some(Modulo),
+        "std.add" => Some(Plus),
+        "std.sub" => Some(Minus),
+        "std.eq" => Some(Eq),
+        "std.ne" => Some(NotEq),
+        "std.gt" => Some(Gt),
+        "std.lt" => Some(Lt),
+        "std.gte" => Some(GtEq),
+        "std.lte" => Some(LtEq),
+        "std.and" => Some(And),
+        "std.or" => Some(Or),
+        "std.concat" => Some(StringConcat),
```

Comment:
Ideally, this would use `std::STD_X` so we have it "linked" with RQ.

Reply:
Yes, I left some notes [here](https://github.com/PRQL/prql/pull/2532/files#diff-2540aa8b42227cb2ab1983dcc56db5f2401778a2ba89fb1c33525c31c59d7e6bR82-R87), I agree the current state isn't ideal. I'm not sure how much we want to have these deep in the compiler with something like an Enum...